### ISSUE TITLE
[BUG] cudf::apply_boolean_mask returns columns with incorrect null_count

### ISSUE THREAD

**Issue by jeanp413** [rapidsai/cudf#6733]
**Describe the bug**
`cudf::apply_boolean_mask` returns columns with incorrect null_count

**Steps/Code to reproduce bug**
Run this unit test
```c++
TEST_F(ApplyBooleanMask, bug)
{
  cudf::size_type inputRows = 75000;

  auto seq1 = cudf::test::make_counting_transform_iterator(0, [](auto i) { return i; });
  auto valid_seq1 = cudf::test::make_counting_transform_iterator(0, [](auto row) { return true; });
  cudf::test::fixed_width_column_wrapper<int64_t, typename decltype(seq1)::value_type> col1(seq1, seq1 + inputRows, valid_seq1);

  cudf::table_view input{{col1}};

  auto seq3 = cudf::test::make_counting_transform_iterator(0, [](auto i) {
    return (i % 277) == 0;
  });
  cudf::test::fixed_width_column_wrapper<bool> boolean_mask(seq3, seq3 + inputRows);

  auto got = cudf::apply_boolean_mask(input, boolean_mask);
  auto out_col = got->get_column(0).view();
  auto expected_null_count = cudf::count_unset_bits(out_col.null_mask(), 0, out_col.size());

  ASSERT_EQ(out_col.null_count(), expected_null_count);
}
```

**Expected behavior**
`cudf::apply_boolean_mask` should return a table containing columns with the correct null_count

**Comment by harrism** [comment]
@jeanp413 I am unable to reproduce this locally. I added your gtest (thanks!) to `apply_boolean_mask_tests.cpp` and recompiled. Here's the output: 

```
(rapids) rapids@compose:~/cudf/cpp/build/release$ gtests/STREAM_COMPACTION_TEST --gtest_filter=*Issue6733*
Note: Google Test filter = *Issue6733*
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from ApplyBooleanMask
[ RUN      ] ApplyBooleanMask.Issue6733
[       OK ] ApplyBooleanMask.Issue6733 (1 ms)
[----------] 1 test from ApplyBooleanMask (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (1 ms total)
[  PASSED  ] 1 test.
```

I'm running on a Quadro GV100 (Volta) on CUDA 11.0, Ubuntu 18.04, gcc 8.4.0.

Can you add your system details?

**Comment by jeanp413** [comment]
@harrism I update the first post with my environment info
I'm running locally on my laptop GPU so it could be the fewer number of SM that cause it to fail
`GPU 0: GeForce GTX 1050 Ti with Max-Q Design (UUID: GPU-c896c1e4-38a3-964f-a366-30fd7ff92b1d)`

**Comment by harrism** [comment]
I don't have access to a Pascal GPU...

**Comment by williamBlazing** [comment]
@harrism This issue is one that we have seen sporadically showing up in our tests for several months (it would only ever show up in one particular query out of hundreds), but we have never been able to reproduce it with any degree of repeatability until this unit test. When it did happen it was usually locally, and not in gpuCI, which means its more likely to happen in certain hardware. 

Although, I would not call this a blocker, its certainly a worry, since when the issue occurs it can output incorrect results in a query. There is no workaround that we have found, nor any way to detect when it happens, except validating the results.

**Comment by harrism** [comment]
Sounds like a heisenbug...  Going to bump it to 0.18.  Does it repro reliably on the GTX 1050  @jeanp413 ?

**Comment by jeanp413** [comment]
yeah I can repro 100%

**Comment by harrism** [comment]
@jeanp413 I added this test in #6903 and I have now tested it on a P100 GPU (the only Pascal I have access to). I ran it in a loop 1000 times and it never failed.

So it must be specific to smaller Pascal GPUs. :(

**Comment by nvdbaranec** [comment]
Necromancy:     This appears to be a real bug masked by a second bug (https://github.com/rapidsai/cudf/issues/7210)

If the null count check is re-enabled, this bug can be reproduced (although you have to crank the size up to 82,000 for some reason).  There's a similar check failing in `drop_nulls` the same way. The culprit appears to be the scatter kernel in `copy_if`

**Comment by harrism** [comment]
Turns out the size at which it fails depends on the GPU you are running on. I have fixed it in #7219
